security: add CA and client certs for tenant usage

tbg opened the pull request:
This commit introduces methods that create a CA certificate for use with
the auth broker envisioned in #49105 and, from that CA, client
certificates for use in #47898. They are not exposed through the cli
(yet), though the certificate loader already supports them.

Touches #49105.
Touches #47898.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49864)
<!-- Reviewable:end -->

knz reviewed (approved):
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @bdarnell)



<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @tbg)

___
*[pkg/security/certs.go, line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94N27REkTqBTiqnl_C:-M94N27REkTqBTiqnl_D:byh9414) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certs.go#L505)):*
> ```Go
> }
> 
> // WriteTenantClientPair writes a TenantClientPair into certsDir.
> ```

Where (on which kind of server) do tenant certs get written to disk? In my understanding of the design these certs are all ephemeral and can be kept in memory. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @bdarnell)

___
*[pkg/security/certs.go, line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94N27REkTqBTiqnl_C:-M94_K79A8eQQ69hVTWx:b-46hbqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certs.go#L505)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Where (on which kind of server) do tenant certs get written to disk? In my understanding of the design these certs are all ephemeral and can be kept in memory. 
</blockquote></details>

Your understanding is correct. In production a client pair either comes from the auth broker (either directly or injected at pod startup, likely through a file then?). I could move this into `_test.go` to keep the public API clean for now, though I anticipated using this for a first plumbing pass of the security work that switches to these certificates (without using the auth broker yet) as well.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> Reviewed 4 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @tbg)

___
*[pkg/security/certificate_loader.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94kyV6BuTx-BKi-fIh:-M94kyV6BuTx-BKi-fIi:b-rr2t57) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certificate_loader.go#L90)):*
> ```Go
> 	ClientPem
> 	// TenantClientPem describes a SQL tenant client certificate.
> 	TenantClientPem
> ```

How does TenantClientPem relate to NodePem (the "node client" cert)? Wouldn't a SQL tenant pod *always* use a tenant cert instead of a node cert? I'm tempted to say we shouldn't have separate variables for TenantClientPem and should just load the tenant cert *as* the node client cert. 

At first I wanted to make this argument for TenantCAPem as well, but this gets trickier - KV servers need to be able to accept connections from both the system tenant and other tenants, and they must be able to distinguish between them (since the tenant CA is online in the auth broker, it's important that it not be able to mint omnipotent node certs). So maybe if we keep that we might as well keep TenantClientPem too. 
___
*[pkg/security/certificate_manager.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94m2lECHJcjnIiiOk0:-M94m2lECHJcjnIiiOk1:b-osgklk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certificate_manager.go#L198)):*
> ```Go
> func CACertFilename() string { return "ca" + certExtension }
> 
> // TenantCACertPath returns the expected file path for the Tenant CA
> ```

FYI I have a branch I'm working on that unexports all the "Path" methods from CertificateManager. 
___
*[pkg/security/certificate_manager.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94mS9N9BX3MGr0lwMW:-M94mS9N9BX3MGr0lwMX:bsc6tp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certificate_manager.go#L269)):*
> ```Go
> // TenantClientKeyFilename returns the expected file name for the user's key.
> func TenantClientKeyFilename(tenantIdentifier string) string {
> 	return "tenant." + tenantIdentifier + keyExtension
> ```

This makes me nervous because it implies we might have keys for multiple clients written to disk in the same place. As I said in certs.go, I'd just avoid doing anything with naming conventions for these certs. 
___
*[pkg/security/certs.go, line 443 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94iUhb3TqLgfV_ZO_E:-M94iUhb3TqLgfV_ZO_F:b54zw03) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certs.go#L443)):*
> ```Go
> }
> 
> // TenantClientPair are client certs for use with multi-tenancy.
> ```

Why a specific type instead of `KeyPair`? Why hard-code RSA for the private key ?(instead of `crypto.PrivateKey`. For that matter, we should get @aaron-crl 's opinion on whether we should be using RSA, ECDSA, or something else here. This is a purely internal part of the system so we don't have to worry as much about rest-of-world compatibility) What format are those bytes in? Why isn't this just a `tls.Certificate` (which despite the name can carry a private key)?

(for a lot of my comments the answer could be "because that's what this file already does elsewhere". It's accumulated a lot of tech debt but we don't have to fix it all right now)
___
*[pkg/security/certs.go, line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94N27REkTqBTiqnl_C:-M94bo6zBIer98RtkKYz:bylp18q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certs.go#L505)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Your understanding is correct. In production a client pair either comes from the auth broker (either directly or injected at pod startup, likely through a file then?). I could move this into `_test.go` to keep the public API clean for now, though I anticipated using this for a first plumbing pass of the security work that switches to these certificates (without using the auth broker yet) as well.
</blockquote></details>

OK. For some additional context, I'm working on some cleanups here to make the certificate manager less file-centric (anticipating future usage of HSM/KMS), and to split it up so we don't have all the different use cases sharing one cert manager object with a bunch of possibly-nil fields. 

So I don't think following the current cert patterns is a requirement or even necessarily a good idea. If you'll just be loading certs from disk in testing scenarios, don't worry about enforcing conventions and require the path to the  files to be passed on the command line. 
___
*[pkg/security/x509.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94qHfxErajjD6bIgES:-M94qHfxErajjD6bIgET:bakxsc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/x509.go#L198)):*
> ```Go
> 	lifetime time.Duration,
> 	user string,
> 	hosts ...string,
> ```

Why would client certs have hosts? Are you trying to make them non-transferable and restricted to a single source pod? Does that work? It kind of makes sense but I can't find any reference to this being done. (and in particular verifying clients with DNS lookups is problematic; it might only make sense for it to use IP addresses)


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)

___
*[pkg/security/certificate_loader.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94kyV6BuTx-BKi-fIh:-M94u8UW5qXsA3kkhJSA:b-bd0t76) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certificate_loader.go#L90)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

How does TenantClientPem relate to NodePem (the "node client" cert)? Wouldn't a SQL tenant pod *always* use a tenant cert instead of a node cert? I'm tempted to say we shouldn't have separate variables for TenantClientPem and should just load the tenant cert *as* the node client cert. 

At first I wanted to make this argument for TenantCAPem as well, but this gets trickier - KV servers need to be able to accept connections from both the system tenant and other tenants, and they must be able to distinguish between them (since the tenant CA is online in the auth broker, it's important that it not be able to mint omnipotent node certs). So maybe if we keep that we might as well keep TenantClientPem too. 
</blockquote></details>

Yes, tenants have no use for node certs. They use their tenant client cert to talk to KV and they also have some additional cert they use for their Postgres-TLS which doesn't factor in here (yet, though we'll need one). Effectively, we'll set up an `rpc.Context` for them that uses the tenant client cert instead of the node client cert but otherwise looks mostly the same.

For practical reasons, I would prefer to just "do how this package does" at the time of writing and (i.e. this PR mostly as is) - if you're amenable to it - rely on you to revisit when you're moving away from the file-centric view (which implies that you'll be figuring out the certs-dir complications across the codebase that would otherwise be front-loaded on me).
I'm mostly trying to move through quickly to get these certs into the right places in the SQL tenant cli using the standard `--certs-dir` flag to build out the tenant server endpoint in the KV layer, and as of this PR I'm there (mod the auth broker, i.e. for testing with filesystem certs, and possibly multiple different tenant's processes at once). If I do anything differently, I have to rework the flags, how the RPC context loads the certs, etc, I think there's just going to be a lot of fallout that I'm not the right person for handling; I think this is better addressed in the work that you've set to do, if I understand it correctly.

I also think there might be a good case for loading tenant cli certs from disk in prod: we will need to inject an initial tenant cert at boot time; naively the filesystem is the right way to do it (or is an env var better?).

Let me know which changes you want me to make before merging.
___
*[pkg/security/certificate_manager.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94mS9N9BX3MGr0lwMW:-M952dVh88PKI-fVLDdK:bssmxf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certificate_manager.go#L269)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This makes me nervous because it implies we might have keys for multiple clients written to disk in the same place. As I said in certs.go, I'd just avoid doing anything with naming conventions for these certs. 
</blockquote></details>

:thumbsup: but see my other comment.
___
*[pkg/security/certs.go, line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94N27REkTqBTiqnl_C:-M95389s8TyCWop5LE2E:bux9rgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/certs.go#L505)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

OK. For some additional context, I'm working on some cleanups here to make the certificate manager less file-centric (anticipating future usage of HSM/KMS), and to split it up so we don't have all the different use cases sharing one cert manager object with a bunch of possibly-nil fields. 

So I don't think following the current cert patterns is a requirement or even necessarily a good idea. If you'll just be loading certs from disk in testing scenarios, don't worry about enforcing conventions and require the path to the  files to be passed on the command line. 
</blockquote></details>

I agree in principal, but see my other comment.
___
*[pkg/security/x509.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94qHfxErajjD6bIgES:-M94tqxc4Y_Xv7S4M-k7:bfszye4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/x509.go#L198)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why would client certs have hosts? Are you trying to make them non-transferable and restricted to a single source pod? Does that work? It kind of makes sense but I can't find any reference to this being done. (and in particular verifying clients with DNS lookups is problematic; it might only make sense for it to use IP addresses)
</blockquote></details>

I just saw that it seemed possible, and so it seemed desirable to try to do it. If you consider this exotic/not worth it I can take it out (otherwise I'd just see whether it works once I have the auth broker). It does seem nice to not have the "token" be transferable to another host. OTOH maybe this is over the top and just causes headaches with pod migrations etc if that is a thing.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
No changes required at this time except removing the host option when generating client certs. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-:-M9Jxfa74fwlm47nXAZA:b3j9jr6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)

___
*[pkg/security/x509.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-M94qHfxErajjD6bIgES:-M9JslGK4rlmkb0y3ZCx:b-8jjpyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad18650fe01467157d3fa3c3e449b750070908a0/pkg/security/x509.go#L198)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I just saw that it seemed possible, and so it seemed desirable to try to do it. If you consider this exotic/not worth it I can take it out (otherwise I'd just see whether it works once I have the auth broker). It does seem nice to not have the "token" be transferable to another host. OTOH maybe this is over the top and just causes headaches with pod migrations etc if that is a thing.
</blockquote></details>

I'd leave it out - in our standard setup, we already have client certs with IPs and hostnames set in SANs but they're not used for client cert verification (this is because our node certs are hybrid client/server).  We wouldn't want to add enforcement that would suddenly change the behavior of existing certs. 


<!-- Sent from Reviewable.io -->

tbg committed:
security: add CA and client certs for tenant usage

This commit introduces methods that create a CA certificate for use with
the auth broker envisioned in #49105 and, from that CA, client
certificates for use in #47898. They are not exposed through the cli
(yet), though the certificate loader already supports them.

Touches #49105.
Touches #47898.

Release note: None

tbg reviewed (commented):
TFTR! I removed the host option.

bors r=bdarnell

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49864#-:-M9S8vEEAAtq-LCIPI6S:b7u3nfw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, @knz, and @tbg)



<!-- Sent from Reviewable.io -->
